File: hht-alpha-integrator/environment/app/integrator.py
"""
HHT-alpha (Generalized-alpha) time integrator for constrained multibody systems.

Solves the index-3 DAE system:
    M * a + Phi_q(q)^T * lambda = F(t, q, v)
    Phi(q) = 0

using Newmark-beta time discretization with generalized-alpha weighting.

The model object must provide:
    - mass_matrix() -> (n_q, n_q) array
    - forces(t, q, v) -> (n_q,) array
    - constraints(q) -> (n_c,) array
    - constraint_jacobian(q) -> (n_c, n_q) array
    - gamma(q, v) -> (n_c,) array
    - n_q: int (number of generalized coordinates)
    - n_c: int (number of constraints)

References:
    Chung & Hulbert (1993), "A Time Integration Algorithm for Structural
    Dynamics With Improved Numerical Dissipation: The Generalized-alpha Method"

    Arnold & Bruls (2007), "Convergence of the generalized-alpha scheme for
    constrained mechanical systems"
"""

import numpy as np


class HHTAlphaIntegrator:
    """
    Generalized-alpha integrator for index-3 DAEs in multibody dynamics.

    Given spectral radius at infinity rho_inf in [0, 1]:
      - rho_inf = 1.0: no numerical dissipation (trapezoidal rule)
      - rho_inf = 0.0: maximum numerical dissipation
    """

    def __init__(self, model, rho_inf=0.9, h=0.001,
                 newton_tol=1e-10, newton_max_iter=50):
        """
        Parameters:
            model: Multibody model object (see module docstring for interface)
            rho_inf: Spectral radius at infinity in [0, 1]
            h: Time step size
            newton_tol: Newton-Raphson convergence tolerance
            newton_max_iter: Maximum Newton-Raphson iterations per step
        """
        self.model = model
        self.h = h
        self.newton_tol = newton_tol
        self.newton_max_iter = newton_max_iter

        # TODO: Compute the four generalized-alpha parameters from rho_inf.
        #
        # The parameters alpha_m, alpha_f, beta (Newmark), gamma (Newmark) must
        # satisfy second-order accuracy and unconditional stability conditions
        # for the generalized-alpha family of integrators.
        #
        # Store them as:
        #   self.alpha_m
        #   self.alpha_f
        #   self.beta_N   (Newmark beta)
        #   self.gamma_N  (Newmark gamma)
        raise NotImplementedError("Set generalized-alpha parameters from rho_inf")

    def compute_initial_accelerations(self, q0, v0):
        """
        Compute consistent initial accelerations and Lagrange multipliers
        by solving the constrained equations of motion at t=0.

        The augmented system combines the dynamics equation and the
        acceleration-level constraint equation into a single linear solve.

        Returns: (a0, lambda0) — both numpy arrays
        """
        # TODO: Assemble and solve the augmented linear system for the
        # initial accelerations a0 and Lagrange multipliers lambda0.
        raise NotImplementedError

    def step(self, t, q, v, a, lam):
        """
        Advance the solution by one time step using the generalized-alpha method.

        Uses Newmark formulas:
            q_{n+1} = q_n + h*v_n + h^2*((0.5 - beta)*a_n + beta*a_{n+1})
            v_{n+1} = v_n + h*((1 - gamma)*a_n + gamma*a_{n+1})

        with alpha-weighted residual evaluation and Newton-Raphson iteration
        to solve the resulting nonlinear system for a_{n+1} and lambda_{n+1}.

        The position constraint Phi(q_{n+1}) = 0 is enforced at q_{n+1}.
        Pay careful attention to which state (q_{n+1} vs the alpha-weighted
        interpolant) is used when evaluating each term in the dynamics
        residual — this choice is critical for the convergence order of
        the method when applied to index-3 DAEs.

        Parameters:
            t: current time
            q: positions (n_q,)
            v: velocities (n_q,)
            a: accelerations (n_q,)
            lam: Lagrange multipliers (n_c,)

        Returns: (t_new, q_new, v_new, a_new, lam_new)
        """
        # TODO: Implement the generalized-alpha step:
        #
        # 1. Initialize a_new (predictor) and lam_new
        # 2. Newton-Raphson loop:
        #    a. Compute q_new, v_new from a_new via Newmark formulas
        #    b. Form alpha-weighted quantities for the dynamics residual
        #    c. Evaluate the dynamic residual and position-level constraint
        #    d. Check convergence
        #    e. Assemble the tangent (Jacobian) matrix of the augmented system
        #    f. Solve for corrections and update a_new, lam_new
        # 3. Compute final q_new, v_new after convergence
        raise NotImplementedError

    def simulate(self, t_end, q0, v0, output_interval=1):
        """
        Run the simulation from t=0 to t_end.

        Parameters:
            t_end: final time
            q0: initial positions (n_q,)
            v0: initial velocities (n_q,)
            output_interval: store results every N steps

        Returns: dict with keys
            'time':                  (N,) float array
            'positions':             (N, n_q) float array
            'velocities':            (N, n_q) float array
            'constraint_violations': (N,) float array of ||Phi(q)||
        """
        # TODO: Initialize (compute consistent initial accelerations),
        # time-step loop, record results at the requested interval.
        raise NotImplementedError

File: hht-alpha-integrator/environment/sim/src/integrator.cpp
/*
 * Time integrator for constrained multibody systems modeled as index-3 DAEs.
 *
 * Solves:   M * a  +  Phi_q(q)^T * lambda  =  F(t, q, v)
 *           Phi(q) = 0
 *
 * using an implicit Newmark-family scheme with generalized-alpha weighting
 * for controllable numerical dissipation.
 */

#include "integrator.h"
#include <fstream>
#include <iostream>
#include <cmath>

Integrator::Integrator(DoublePendulum& model, double rho_inf, double h,
                       double newton_tol, int max_iter)
    : model_(model), h_(h), newton_tol_(newton_tol), max_iter_(max_iter)
{
    alpha_m_ = (2.0 * rho_inf - 1.0) / (rho_inf + 1.0);
    alpha_f_ = rho_inf / (rho_inf + 1.0);
    gamma_N_ = 0.5 - alpha_m_ + alpha_f_;
    beta_N_  = 0.25 * (gamma_N_ + 0.5) * (gamma_N_ + 0.5);
}

void Integrator::computeInitialAccelerations(
    const Eigen::VectorXd& q0, const Eigen::VectorXd& v0,
    Eigen::VectorXd& a0, Eigen::VectorXd& lam0)
{
    int nq = model_.nq();
    int nc = model_.nc();
    Eigen::MatrixXd M   = model_.massMatrix();
    Eigen::VectorXd F   = model_.forces(0.0, q0, v0);
    Eigen::MatrixXd Cq  = model_.constraintJacobian(q0);
    Eigen::VectorXd gam = model_.gamma(q0, v0);

    Eigen::MatrixXd A = Eigen::MatrixXd::Zero(nq + nc, nq + nc);
    A.topLeftCorner(nq, nq) = M;
    A.topRightCorner(nq, nc) = Cq.transpose();
    A.bottomLeftCorner(nc, nq) = Cq;

    Eigen::VectorXd rhs = Eigen::VectorXd::Zero(nq + nc);
    rhs.head(nq) = F;
    rhs.tail(nc) = gam;

    Eigen::VectorXd sol = A.fullPivLu().solve(rhs);
    a0   = sol.head(nq);
    lam0 = sol.tail(nc);
}

void Integrator::step(double& t, Eigen::VectorXd& q, Eigen::VectorXd& v,
                      Eigen::VectorXd& a, Eigen::VectorXd& lam)
{
    int nq = model_.nq();
    int nc = model_.nc();
    double h    = h_;
    double beta = beta_N_;
    double gN   = gamma_N_;
    double am   = alpha_m_;
    double af   = alpha_f_;
    Eigen::MatrixXd M = model_.massMatrix();

    Eigen::VectorXd a_new   = a;
    Eigen::VectorXd lam_new = lam;
    Eigen::VectorXd q_old   = q;
    Eigen::VectorXd v_old   = v;

    for (int k = 0; k < max_iter_; k++) {
        // Newmark update formulas
        Eigen::VectorXd q_new = q_old + h * v_old
                                + h*h * ((0.5 - beta) * a + beta * a_new);
        Eigen::VectorXd v_new = v_old + h * ((1.0 - gN) * a + gN * a_new);

        // Weighted quantities for the alpha scheme
        Eigen::VectorXd a_am = (1.0 - am) * a_new + am * a;
        Eigen::VectorXd q_af = (1.0 - af) * q_new + af * q_old;
        Eigen::VectorXd v_af = (1.0 - af) * v_new + af * v_old;
        double t_af = t + (1.0 - af) * h;

        // Model evaluation
        Eigen::VectorXd Phi       = model_.constraints(q_new);
        Eigen::MatrixXd Phi_q_dyn = model_.constraintJacobian(q_new);
        Eigen::MatrixXd Phi_q_con = model_.constraintJacobian(q_new);
        Eigen::VectorXd F_af      = model_.forces(t_af, q_af, v_af);

        // Assemble residual
        Eigen::VectorXd R_dyn = M * a_am
                                + Phi_q_dyn.transpose() * lam_new - F_af;
        Eigen::VectorXd R(nq + nc);
        R.head(nq) = R_dyn;
        R.tail(nc) = Phi;

        if (R.norm() < newton_tol_) break;

        // Assemble tangent matrix
        Eigen::MatrixXd J = Eigen::MatrixXd::Zero(nq + nc, nq + nc);
        J.topLeftCorner(nq, nq) = M;
        J.topRightCorner(nq, nc) = Phi_q_dyn.transpose();
        J.bottomLeftCorner(nc, nq) = h*h * beta * Phi_q_con;

        Eigen::VectorXd delta = J.fullPivLu().solve(-R);
        a_new   += delta.head(nq);
        lam_new += delta.tail(nc);
    }

    q = q_old + h * v_old + h*h * ((0.5 - beta) * a + beta * a_new);
    v = v_old + h * ((1.0 - gN) * a + gN * a_new);
    a   = a_new;
    lam = lam_new;
    t  += h;
}

SimResult Integrator::simulate(double t_end, Eigen::VectorXd q0,
                                Eigen::VectorXd v0, int output_interval)
{
    SimResult result;
    Eigen::VectorXd a0, lam0;
    computeInitialAccelerations(q0, v0, a0, lam0);

    double t = 0.0;
    Eigen::VectorXd q = q0, v = v0, a = a0, lam = lam0;

    result.time.push_back(t);
    result.positions.push_back(q);
    result.velocities.push_back(v);
    result.constraint_violations.push_back(model_.constraints(q).norm());
    result.energies.push_back(model_.totalEnergy(q, v));

    int n_steps = static_cast<int>(std::round(t_end / h_));
    for (int i = 1; i <= n_steps; i++) {
        step(t, q, v, a, lam);
        if (i % output_interval == 0) {
            result.time.push_back(t);
            result.positions.push_back(q);
            result.velocities.push_back(v);
            result.constraint_violations.push_back(
                model_.constraints(q).norm());
            result.energies.push_back(model_.totalEnergy(q, v));
        }
    }
    return result;
}

void Integrator::writeCSV(const std::string& filename,
                           const SimResult& result)
{
    std::ofstream ofs(filename);
    ofs << "time,x1,y1,theta1,x2,y2,theta2,"
        << "vx1,vy1,omega1,vx2,vy2,omega2,"
        << "constraint_violation,total_energy\n";
    ofs << std::scientific;
    ofs.precision(12);
    for (size_t i = 0; i < result.time.size(); i++) {
        ofs << result.time[i];
        for (int j = 0; j < 6; j++)
            ofs << "," << result.positions[i](j);
        for (int j = 0; j < 6; j++)
            ofs << "," << result.velocities[i](j);
        ofs << "," << result.constraint_violations[i];
        ofs << "," << result.energies[i];
        ofs << "\n";
    }
    ofs.close();
}
